### ISSUE TITLE
[SUGGESTION]  Force using non-deprecated code

### ISSUE THREAD

**Issue by define-private-public** [hsutter/cppfront#145]
I've not see the [deprecated](https://en.cppreference.com/w/cpp/language/attributes/deprecated) attribute used to much in APIs, but I have seen my share of programmers who are unknowingly using deprecated code, or at least knowingly being lazy and not updating deprecated API calls.

**I'd like to discuss forcing using non-deprecated code in cppfront**.

I.e., if you used a piece of code that's marked as `deprecated`, compilation should fail.  I think this would lead to people acting better at making sure their code is up to date.  I've also seen the case in other software where one library, which uses deprecated code, can hold back an entire product from being updated.

**Will your feature suggestion eliminate X% of security vulnerabilities of a given kind in current C++ code?**

Not too sure on this.  It's possible that older security algorithms (e.g. hashes) can become insecure over time, thus need to be deprecated and eventually removed.

**Will your feature suggestion eliminate X% of current C++ guidance literature?"**

I don't think so.

**Describe alternatives you've considered.**

I'm not sure what would be a valid alternative here, other than "allow for deprecated code".  Deprecated code could be allowed with an explicit compiler flag, but it should then still show a warning when a deprecated call is compiled.

**Comment by hsutter** [comment]
Thanks!

Note that `[[deprecated]]` is about transition, not primarily about safety. (Yes, many unsafe things get deprecated, but lots of deprecated things are deprecated for reasons other than safety.)

The purpose of `[[deprecated]]` is to enable a gradual transition path, so that a library can notify its users that a change is coming but the users can defer implementing the changes to when they can absorb them in their various project cycles. But `[[deprecated]]` is not intended to be used as a cudgel for library authors to force source changes into all projects that (transitively) use the library; they can already achieve that by just removing the deprecated function outright, and of course that would impact their new library's adoption. So they need a transition path to warn users that an API might not exist in the future, and to give them time to get their projects ready for when that API goes away in the future... so they mark it `[[deprecated]]` with the deliberate intention _not_ to break the client's build.

I'm fine with compilers continuing to provide a "turn `[[deprecated]]` from a warning to an error" switch, but it does need to be opt-in somehow... the programmer needs to be able to control at least whether it's a warning or an error, and additionally probably wants some granularity control on that. For now I think that existing "warnings as errors" switches cover this adequately, and designing a system with more granularity is out of scope for Cpp2 at the moment, as right now the focus is on completing implementation of the basic language.

Some good news though: Once the basic language implementation is complete and fairly solid, I do aim to get to parts of the Cpp2 design that I haven't talked about yet, and one of those is first-class support for versioning... not limited to only deprecation, but also exploring what that means for adding functions and data members to classes, including distinguishing between ABI-compatible changes ("minor" versions in my view) and ABI-breaking changes ("major" versions) and compile-time language support for committing to the version of a type you intend to use. That would reopen this topic in a broader way that I think could address it better than with just deprecation in isolation.

Thanks!

**Comment by filipsajdak** [comment]
Looking forward for versioning topic. Looks promising!
